### PR TITLE
Adding `StickBreakingWeights` distribution class

### PR DESCRIPTION
**Pull request by larryshamalama**
This pull request adds a class for truncated stick-breaking weights which are vital in the construction of Dirichlet Processes, which I hope to work on soon. See section 1.1 in [Ishwaran and James (2001)](http://people.ee.duke.edu/~lcarin/Yuting3.3.06.pdf). My previous attempt as a PR is #4971 .

Some tests are failing, will have to resolve them first.

CC: @ricardoV94 @AustinRochford @fonnesbeck

### CODE REVIEW

File: pymc/distributions/multivariate.py [hunk 1]
```diff
@@ -2136,3 +2137,117 @@ def logp(value, mu, W, alpha, tau):
...
+    def logp(value, alpha, K):
+        if at.lt(K, 0).eval():
+            raise ValueError("K needs to be a positive integer.")
+
+        if at.lt(alpha, 0).eval():
+            raise ValueError("alpha needs to be a positive.")
+
+        if not at.allclose(value.sum(axis=-1), 1).eval():
+            raise ValueError("weights need to sum to 1.")
+
```
Comment (larryshamalama) [inline pymc/distributions/multivariate.py:2317]: Generally, I am using `.eval()` for some checks here, but I don't know if this is generalizable especially that inputs can be `np.ndarray` or Aesara tensors. Is there a better to do these checks?

File: pymc/distributions/multivariate.py [hunk 2]
```diff
@@ -2136,3 +2137,117 @@ def logp(value, mu, W, alpha, tau):
...
+    def logp(value, alpha, K):
+        if at.lt(K, 0).eval():
+            raise ValueError("K needs to be a positive integer.")
+
+        if at.lt(alpha, 0).eval():
+            raise ValueError("alpha needs to be a positive.")
+
+        if not at.allclose(value.sum(axis=-1), 1).eval():
+            raise ValueError("weights need to sum to 1.")
```
Comment (ricardoV94) [inline pymc/distributions/multivariate.py]: These should just be conditions in bound, and not `.eval` preemptively (you would need to know the inputs to be able to evaluate them)

File: pymc/tests/test_distributions.py [hunk 3]
```diff
@@ -2136,6 +2137,42 @@ def test_dirichlet_2D(self):
...
+    def test_stickbreakingweights(self, value, alpha, K, logp):
+        with Model() as model:
+            StickBreakingWeights("sbw", alpha=alpha, K=K)
+        pt = {"sbw": value}
+        assert_almost_equal(
+            model.fastlogp(pt),
+            logp,
+            decimal=select_by_precision(float64=6, float32=2),
+            err_msg=str(pt),
+        )
```
Comment (larryshamalama) [inline pymc/tests/test_distributions.py:2151]: I drew inspiration `test_ex_gaussian` because there is no scipy or numpy logp equivalent. Here is the error that I'm getting with this test.

Screen Shot 2021-11-17 at 5 42 05 PM

File: pymc/tests/test_distributions_random.py [hunk 4]
```diff
@@ -1175,6 +1175,26 @@ class TestDirichlet(BaseTestDistribution):
...
+            20,
+        ),
+        (11, 5, 20),
+        (3, 13, 5, 20),
+    ]
+    tests_to_run = [
+        "check_pymc_params_match_rv_op",
+        "check_rv_size",
+    ]
+
```
Comment (larryshamalama) [inline pymc/tests/test_distributions_random.py:1199]: This test is passing!

File: pymc/tests/test_distributions_moments.py [hunk 5]
```diff
@@ -897,4 +898,12 @@ def test_asymmetriclaplace_moment(b, kappa, mu, size, expected):
 def test_rice_moment(nu, sigma, size, expected):
     with Model() as model:
         Rice("x", nu=nu, sigma=sigma, size=size)
+
+
+@pytest.mark.parametrize(
+    "alpha, K, size, expected", [(3, 11, None, (3 / 4) ** np.arange(12) * 1 / 3)]
+)
+def test_stickbreakingweights_moment(alpha, K, size, expected):
```
Comment (larryshamalama) [inline pymc/tests/test_distributions_moments.py:1117]: Here, I am obtaining the following error and I am not sure where `result` comes from. Does it represent a random draw? My understanding is that, for other moment tests, `result` is not a random draw. I obtain the same error with the same numbers every time I try to run this test.

Screen Shot 2021-11-17 at 5 44 19 PM

**Review by larryshamalama — COMMENTED** [review]
Two sets of tests not passing locally in `test_distributions.py` and `test_distributions_moments.py`. The test in `test_distributions_random.py` seems okay (from my perspective)

File: pymc/distributions/multivariate.py (hunk 2 above)
Comment (larryshamalama) [reply pymc/distributions/multivariate.py]: Okay, will add these to `bound`, thanks! I just finished writing some comments in a code review but am much slower than your response speed 😅

File: pymc/tests/test_distributions_moments.py (hunk 5 above)
Comment (ricardoV94) [reply pymc/tests/test_distributions_moments.py:1117]: `result` is the `moment`. It seems to deviate slightly from the expected moment. How did you pre-compute it? Are you sure it matches the `get_moment` computation?

Comment (larryshamalama) [reply pymc/tests/test_distributions_moments.py:1117]: Fixed, my `get_moment` and unit tests were wrong

File: pymc/tests/test_distributions.py [hunk 6]
```diff
@@ -2112,6 +2113,32 @@ def test_dirichlet(self, n):
...
+            (np.tile(1, 13) / 13, 2, 12, 13.286898065112881),
+            (np.array([0.001] * 10 + [0.99]), 0.1, 10, -20.66907735582068),
+            (np.append(0.5 ** np.arange(1, 20), 0.5 ** 20), 5, 19, 92.59518981534681),
+        ],
+    )
+    def test_stickbreakingweights(self, value, alpha, K, logp):
+        with Model() as model:
+            StickBreakingWeights("sbw", alpha=alpha, K=K)
+        pt = {"sbw": value}
+        assert_almost_equal(
```
Comment (larryshamalama) [inline pymc/tests/test_distributions.py:2146]: I am getting a `TypeError: Unknown input or state: sbw.` error here... Am I supposed to name things different or replace `pt` by `pt = {"sbw_simplex__": value}`? If I do that, then I obtained `AssertionError` but I'm not sure if that's the correct approach...

**Review by larryshamalama — COMMENTED** [review]
Question about `test_distributions.py`

File: pymc/tests/test_distributions.py (hunk 6 above)
Comment (ricardoV94) [reply pymc/tests/test_distributions.py:2146]: You have to disable transform, with `transform=None`, or else your reference logp has to account for the Jacobian term as well.

Comment (larryshamalama) [reply pymc/tests/test_distributions.py:2146]: This works well, thanks!

File: pymc/distributions/multivariate.py (hunk 2 above)
Comment (larryshamalama) [reply pymc/distributions/multivariate.py]: Quick comment that I replace `bound` with `check_parameters` following a recent merged PR (to my understanding)

File: pymc/tests/test_distributions.py (hunk 3 above)
Comment (larryshamalama) [reply pymc/tests/test_distributions.py:2151]: Solved by using `transform=None`, which is okay when we're just testing the logp method

File: pymc/distributions/multivariate.py [hunk 7]
```diff
@@ -2153,3 +2160,120 @@ def logp(value, mu, W, alpha, tau):
...
+
+
+class StickBreakingWeightsRV(RandomVariable):
+    name = "stick_breaking_weights"
+    ndim_supp = 1
+    ndims_params = [0, 0]
+    dtype = "floatX"
+    _print_name = ("StickBreakingWeights", "\\operatorname{StickBreakingWeights}")
+
+    def __call__(self, alpha=1.0, K=10, size=None, **kwargs):
```
Comment (ricardoV94) [inline pymc/distributions/multivariate.py]: Are you sure you want default parameters? Any reason to think these are particularly popular?

File: pymc/distributions/multivariate.py [hunk 8]
```diff
@@ -2153,3 +2160,120 @@ def logp(value, mu, W, alpha, tau):
...
+    def _infer_shape(self, size, dist_params, param_shapes=None):
+        alpha, K = dist_params
+
+        size = tuple(size)
+
+        return size + (K + 1,)
+
+    @classmethod
+    def rng_fn(cls, rng, alpha, K, size):
+        if K < 0:
```
Comment (ricardoV94) [inline pymc/distributions/multivariate.py:2208]: I would remove the "integer" from the message, it makes it less clear the problem is the being negative

File: pymc/distributions/multivariate.py [hunk 9]
```diff
@@ -2153,3 +2160,120 @@ def logp(value, mu, W, alpha, tau):
...
+    @classmethod
+    def dist(cls, alpha, K, *args, **kwargs):
+        alpha = at.as_tensor_variable(floatX(alpha))
+        K = at.as_tensor_variable(intX(K))
+
+        if alpha.ndim > 0:
+            raise ValueError("alpha must be a scalar.")
+
+        if K.ndim > 0:
+            raise ValueError("K must be a scalar.")
```
Comment (ricardoV94) [inline pymc/distributions/multivariate.py]: If you move these checks to make_node in the RV you don't need to repeat them in the rng_fn because you already know that the RV cannot be created with the wrong input types

File: pymc/distributions/multivariate.py [hunk 10]
```diff
@@ -2153,3 +2160,120 @@ def logp(value, mu, W, alpha, tau):
...
+        logp += -(K - 1) * betaln(1, alpha)
+        logp += alpha * at.log(value[..., -1])  # check this
+
+        return check_parameters(
+            logp,
+            alpha > 0,
+            K > 0,
+            at.all(value >= 0),
+            at.all(value <= 1),
+            at.allclose(value.sum(axis=-1), 1),
```
Comment (ricardoV94) [inline pymc/distributions/multivariate.py]: Value checks should now be done outside of check parameters with an explicit switch statement that returns -inf if they are not met.

Also to be more specific you can add an `axis=-1` to the `at.all(value...)` to only switch the logp of the "invalid values"

You can see the multinomial distribution for an example.

File: pymc/distributions/multivariate.py [hunk 11]
```diff
@@ -2153,3 +2160,120 @@ def logp(value, mu, W, alpha, tau):
...
+            at.log(
+                at.cumsum(
+                    value[..., ::-1],
+                    axis=-1,
+                )
+            ),
+            axis=-1,
+        )
+        logp += -(K - 1) * betaln(1, alpha)
+        logp += alpha * at.log(value[..., -1])  # check this
```
Comment (ricardoV94) [inline pymc/distributions/multivariate.py]: Should that comment be removed?

File: pymc/distributions/multivariate.py [hunk 12]
```diff
@@ -2153,3 +2160,120 @@ def logp(value, mu, W, alpha, tau):
...
+
+        return check_parameters(
```
Comment (ricardoV94) [inline pymc/distributions/multivariate.py:2349]: You can add a `msg` kwargs that is shown when the checks are invalid. It also makes the debug_print easier to read. 

There are some examples in most other distributions

File: pymc/tests/test_distributions.py [hunk 13]
```diff
@@ -2185,6 +2186,32 @@ def test_dirichlet_invalid(self):
...
+            model.fastlogp(pt),
+            logp,
+            decimal=select_by_precision(float64=6, float32=2),
+            err_msg=str(pt),
+        )
+
+    @pytest.mark.parametrize("n", [2, 3])
+    def test_multinomial(self, n):
+        self.check_logp(
+            Multinomial, Vector(Nat, n), {"p": Simplex(n), "n": Nat}, multinomial_logpdf
```
Comment (ricardoV94) [inline pymc/tests/test_distributions.py]: This change seems unrelated. If you add a comma after the last function argument, pre-commit won't try to change those lines

File: pymc/distributions/multivariate.py (hunk 12 above)
Comment (ricardoV94) [inline pymc/distributions/multivariate.py:2349]: We need tests for the logp parameter checks and multivariate logp `value`s.

You can check `test_dirichlet_multinomial_vectorized` and `test_dirichlet_multinomial_invalid` for examples of both.

If you add a test that uses `check_logp` it will test for invalid scalar parameters `alpha` and `k` automatically. You would only need manual tests for the `value`

File: pymc/tests/test_distributions_random.py [hunk 14]
```diff
@@ -1163,6 +1163,26 @@ class TestDirichlet(BaseTestDistribution):
     ]
 
 
+class TestStickBreakingWeights(BaseTestDistribution):
```
Comment (ricardoV94) [inline pymc/tests/test_distributions_random.py:1179]: Is there any obvious statistical property you could easily test to be more confident the rng_fn is returning valid samples?

The values must some to 1 right? Maybe other things as well?

**Review by ricardoV94 — COMMENTED** [review]
Looks great. Left some comments bellow, mostly concerning tests. Glad to see this PR entering the polishing phase!

File: pymc/distributions/multivariate.py (hunk 7 above)
Comment (larryshamalama) [reply pymc/distributions/multivariate.py]: Hmm no... they were somewhat arbitrary and perhaps there just to serve as a reminder about the functionality of `__call__`. Would you recommend removing this method?

File: pymc/distributions/multivariate.py (hunk 11 above)
Comment (larryshamalama) [reply pymc/distributions/multivariate.py]: Yes!

File: pymc/tests/test_distributions_random.py (hunk 14 above)
Comment (larryshamalama) [reply pymc/tests/test_distributions_random.py:1179]: Yes values must sum to one. Should I also write a test that checks `0 <= value <= 1`?

File: pymc/tests/test_distributions.py (hunk 13 above)
Comment (larryshamalama) [reply pymc/tests/test_distributions.py]: Oh this might have been dangling from me copy pasting things for visual references and/or mix-ups when rebasing... I will delete this bit

**Review by larryshamalama — COMMENTED** [review]
December 20 work in progress

File: pymc/distributions/multivariate.py [hunk 15]
```diff
@@ -2160,3 +2167,139 @@ def logp(value, mu, W, alpha, tau):
...
+    def make_node(self, rng, size, dtype, alpha, K):
+
+        alpha = at.as_tensor_variable(alpha)
+        K = at.as_tensor_variable(intX(K))
+
+        # if at.lt(K, 0):
+        #     print(at.lt(K, 0).eval())
+        #     print(K.eval() < 0)
+        #     raise ValueError("K needs to be positive.")
+
```
Comment (larryshamalama) [inline pymc/distributions/multivariate.py]: Screen Shot 2021-12-21 at 12 50 29 AM

This check in `make_node` is commented out for now because I keep mistakenly raising a `ValueError`

File: pymc/distributions/multivariate.py [hunk 16]
```diff
@@ -2160,3 +2167,139 @@ def logp(value, mu, W, alpha, tau):
...
+            at.or_(
+                at.any(
+                    at.and_(at.le(value, 0), at.ge(value, 1)),
+                    axis=-1,
+                ),
+                at.or_(
+                    at.bitwise_not(at.allclose(value.sum(-1), 1)),
+                    at.neq(value.shape[-1], K + 1),
+                ),
+            ),
```
Comment (larryshamalama) [inline pymc/distributions/multivariate.py:2344]: There are a lot of `at.or_`... is there a way to combine multiple elements on which we want to do the "or" operation? Would `scan` help us here? (I know nothing about `scan`)

File: pymc/tests/test_distributions.py [hunk 17]
```diff
@@ -2185,6 +2186,34 @@ def test_dirichlet_invalid(self):
...
+        )
+
+    def test_stickbreakingweights_invalid(self):
+        sbw = pm.StickBreakingWeights.dist(3.0, 3)
+        sbw_wrong_K = pm.StickBreakingWeights.dist(3.0, 7)
+        assert pm.logp(sbw, np.array([0.4, 0.3, 0.2, 0.15])).eval() == -np.inf
+        assert pm.logp(sbw, np.array([1.1, 0.3, 0.2, 0.1])).eval() == -np.inf
+        assert pm.logp(sbw, np.array([0.4, 0.3, 0.2, -0.1])).eval() == -np.inf
+        assert pm.logp(sbw_wrong_K, np.array([0.4, 0.3, 0.2, 0.1])).eval() == -np.inf
+
```
Comment (larryshamalama) [inline pymc/tests/test_distributions.py:2160]: Here are tests for invalid logp values akin to `pm.Multinomial`

File: pymc/tests/test_distributions_random.py [hunk 18]
```diff
@@ -1176,6 +1176,40 @@ class TestDirichlet(BaseTestDistribution):
...
+            alpha=3.5,
+            K=19,
+            size=(2, 3, 5),
+            rng=default_rng,
+        )
+
+        assert np.allclose(draws.sum(-1).eval(), 1)
+        assert np.all(draws.eval() >= 0)
+        assert np.all(draws.eval() <= 1)
+
```
Comment (larryshamalama) [inline pymc/tests/test_distributions_random.py:1212]: Another new test here

**Review by larryshamalama — COMMENTED** [review]
Two new sets of tests + question in `make_node`

File: pymc/distributions/multivariate.py [hunk 19]
```diff
@@ -2160,3 +2167,182 @@ def logp(value, mu, W, alpha, tau):
...
+            -np.inf,
+            logp,
+        )
+
+        return check_parameters(
+            logp,
+            alpha > 0,
+            K > 0,
+            msg="alpha > 0, K > 0, 0 <= value <= 1, sum(value) == 1",
```
Comment (ricardoV94) [inline pymc/distributions/multivariate.py]: The value checks shouldn't be part of the message since they happen outside of parameter checks

File: pymc/distributions/multivariate.py (hunk 16 above)
Comment (ricardoV94) [reply pymc/distributions/multivariate.py:2344]: This approach looks reasonable

File: pymc/distributions/multivariate.py (hunk 15 above)
Comment (larryshamalama) [reply pymc/distributions/multivariate.py]: Temporarily I have:

```{Python}
if K.eval() < 0:
            raise ValueError("K needs to be positive.")
```

File: pymc/distributions/multivariate.py [hunk 20]
```diff
@@ -2167,3 +2174,183 @@ def logp(value, mu, W, alpha, tau):
...
+    dtype = "floatX"
+    _print_name = ("StickBreakingWeights", "\\operatorname{StickBreakingWeights}")
+
+    def make_node(self, rng, size, dtype, alpha, K):
+
+        alpha = at.as_tensor_variable(alpha)
+        K = at.as_tensor_variable(intX(K))
+
+        if K.eval() < 0:
+            raise ValueError("K needs to be positive.")
```
Comment (larryshamalama) [inline pymc/distributions/multivariate.py]: If I write `if at.lt(K, 0)`, it often seems to incorrectly evaluate to True... Is there a better way to write this `if` statement?

File: pymc/tests/test_distributions.py [hunk 21]
```diff
@@ -2123,6 +2124,34 @@ def test_dirichlet_invalid(self):
...
+    @pytest.mark.parametrize(
+        "value,alpha,K,logp",
+        [
+            (np.array([5, 4, 3, 2, 1]) / 15, 0.5, 4, 2.2057773112876893),
+            (np.tile(1, 13) / 13, 2, 12, 13.286898065112881),
+            (np.array([0.001] * 10 + [0.99]), 0.1, 10, -20.66907735582068),
+            (np.append(0.5 ** np.arange(1, 20), 0.5 ** 20), 5, 19, 92.59518981534681),
+        ],
+    )
+    def test_stickbreakingweights(self, value, alpha, K, logp):
```
Comment (ricardoV94) [inline pymc/tests/test_distributions.py]:
```suggestion
    def test_stickbreakingweights_logp(self, value, alpha, K, logp):
```

Comment (ricardoV94) [reply pymc/tests/test_distributions.py]: A test with 2d value might be useful

File: pymc/distributions/multivariate.py (hunk 20 above)
Comment (ricardoV94) [reply pymc/distributions/multivariate.py]: Here you can't raise a ValueError yet because the values are possibly still symbolic.  Such error should be in the `rng_fn` where it is already concrete.

In contrast you can raise erros for ndims, because those are fixed.

**Review by ricardoV94 — COMMENTED** [review]
Looks great! Left some comments, let me know if anything is confusing

File: pymc/tests/test_distributions.py (hunk 21 above)
Comment (larryshamalama) [reply pymc/tests/test_distributions.py]: Added!

File: pymc/tests/test_distributions_random.py [hunk 22]
```diff
@@ -1176,6 +1176,41 @@ class TestDirichlet(BaseTestDistribution):
...
+        draws = pm.StickBreakingWeights.dist(
+            alpha=3.5,
+            K=19,
+            size=(2, 3, 5),
+            rng=default_rng,
+        )
+
+        assert np.allclose(draws.sum(-1).eval(), 1)
+        assert np.all(draws.eval() >= 0)
+        assert np.all(draws.eval() <= 1)
```
Comment (ricardoV94) [inline pymc/tests/test_distributions_random.py]: No need to call eval 3 times, you can store the results in a variable and then do the 3 checks on them

Comment (larryshamalama) [reply pymc/tests/test_distributions_random.py]: Are you saying to do something like: `assert all([bool, bool2, bool3])`

Comment (ricardoV94) [inline pymc/tests/test_distributions_random.py]:
```suggestion
        draws = draws.eval()
        assert np.allclose(draws.sum(-1), 1)
        assert np.all(draws >= 0)
        assert np.all(draws <= 1)
```

Comment (larryshamalama) [reply pymc/tests/test_distributions_random.py]: Oops, got it!

Comment (larryshamalama) [reply pymc/tests/test_distributions_random.py]: Thanks! Will commit this asap

**Review by ricardoV94 — APPROVED** [review]
LGTM. Awesome work @larryshamalama

**Comment by larryshamalama** [comment]
As always, thanks so much for all your suggestions and help @ricardoV94

**Comment by twiecki** [comment]
Would be great to have an example NB in pymc-examples that shows how to use this.
